### PR TITLE
[exporter/debug] add ability to set output_paths

### PR DESCRIPTION
**Pull request by ishaish103**
#### Link to tracking issue
Fixes https://github.com/open-telemetry/opentelemetry-collector/issues/10472

### CODE REVIEW

File: exporter/debugexporter/README.md [hunk 1]
```diff
@@ -38,6 +38,7 @@ The following settings are optional:
   Refer to [Zap docs](https://godoc.org/go.uber.org/zap/zapcore#NewSampler) for more details
   on how sampling parameters impact number of messages.
 - `use_internal_logger` (default = `true`): uses the collector's internal logger for output. See [below](#using-the-collectors-internal-logger) for description.
+- `output_paths` (default = `["stdout"]`): a list of URLs or file paths to write logging output to. This option is only used when `use_internal_logger` is `false`. The URLs could only be with "file" schema or without schema. The URLs with "file" schema must be an absolute path. The URLs without schema are treated as local file paths. "stdout" and "stderr" are interpreted as os.Stdout and os.Stderr.
```
Comment (andrzej-stencel) [inline exporter/debugexporter/README.md]: Let's not mention "URLs", it may be confusing to users. I propose to officially support only bare file paths, regardless of what is [internally supported by Zap](https://pkg.go.dev/go.uber.org/zap@v1.27.0#Open).

Comment (andrzej-stencel) [inline exporter/debugexporter/README.md]: I suggest to not use the terms `os.Stdout` and `os.Stderr`. They may be understandable to Go programmers, but not necessarily to all users of the Debug exporter.

```suggestion
- `output_paths` (default = `["stdout"]`): a list of URLs or file paths to write logging output to. This option is only used when `use_internal_logger` is `false`. The URLs could only be with "file" schema or without schema. The URLs with "file" schema must be an absolute path. The URLs without schema are treated as local file paths. Special strings are "stdout" and "stderr".
```

File: exporter/debugexporter/exporter_test.go [hunk 2]
```diff
@@ -124,6 +124,16 @@ func createTestCases() []testCase {
 				return cfg
 			}(),
 		},
+		{
```
Comment (andrzej-stencel) [inline exporter/debugexporter/exporter_test.go:131]: Please add a test case for when user sets `output_paths` but does not set `use_internal_logger` to `false`.

What should happen then? Shouldn't such config be invalid?

Comment (andrzej-stencel) [inline exporter/debugexporter/exporter_test.go:131]: Please add a test case for when user sets `output_paths` to an empty array.

What should happen then? Shouldn't such config be invalid?

Comment (ishaish103) [reply exporter/debugexporter/exporter_test.go:131]: good catch, ill validate it and throw an error if out_paths is null and set_intenral_logger is false

Comment (ishaish103) [reply exporter/debugexporter/exporter_test.go:131]: basically the same thing:
throw an error if out_paths is null and set_intenral_logger is false

File: exporter/debugexporter/README.md (hunk 1 above)
Comment (ishaish103) [reply exporter/debugexporter/README.md]: ok, ill also add validation for the paths as well to validate that its not one of the following:
file:///path/to/file (contains ://)
http://example.com (contains ://)
file:path/to/file (matches scheme: pattern)

**Comment by ishaish103** [comment]
@andrzej-stencel  Hey :) 
ive updated the code, can you please review?

File: exporter/debugexporter/README.md (hunk 1 above)
Comment (andrzej-stencel) [reply exporter/debugexporter/README.md]: Can we remove references to `os.Stdout` and `os.Stderr`? We could link to [Wikipedia](https://en.wikipedia.org/wiki/Standard_streams#Standard_output_(stdout)) if we want to explain what `stdout` and `stderr` stand for.

File: exporter/debugexporter/config.go [hunk 3]
```diff
@@ -47,5 +54,10 @@ func (cfg *Config) Validate() error {
 		return fmt.Errorf("verbosity level %q is not supported", cfg.Verbosity)
 	}
 
+	// If use_internal_logger is false, output_paths must be specified and non-empty
```
Comment (andrzej-stencel) [inline exporter/debugexporter/config.go]: Please alsso add validation for when `output_paths` is specified in config, but `use_internal_logger` is not `false`.

File: exporter/debugexporter/factory.go [hunk 4]
```diff
@@ -128,6 +129,11 @@ func createCustomLogger(exporterConfig *Config) *zap.Logger {
 	encoderConfig.LevelKey = ""
 	// Do not prefix the output with current timestamp.
 	encoderConfig.TimeKey = ""
+	outputPaths := exporterConfig.OutputPaths
+	if len(outputPaths) == 0 {
+		// Default to stdout if not specified
+		outputPaths = []string{"stdout"}
+	}
```
Comment (andrzej-stencel) [inline exporter/debugexporter/factory.go]: This should not be needed? The default is already specified in `createDefaultConfig` function, right?

File: exporter/debugexporter/config.go (hunk 3 above)
Comment (ishaish103) [reply exporter/debugexporter/config.go]: Thanks for the feedback! I've added validation to error when `output_paths` is specified but `use_internal_logger` is true. The validation now checks:

1. If `use_internal_logger` is false, `output_paths` must be specified and non-empty
2. If `use_internal_logger` is true, `output_paths` should not be specified (it's ignored anyway)

I've also updated the default config to have empty `OutputPaths` when `UseInternalLogger` is true, since it's not used in that case. Added a test case to verify this validation works correctly.

File: exporter/debugexporter/factory.go (hunk 4 above)
Comment (ishaish103) [reply exporter/debugexporter/factory.go]: thanks, I've removed the redundant fallback code in `createCustomLogger`. Since:

1. Added validation to error when `output_paths` is specified but `use_internal_logger` is true
2. Removed the redundant fallback in `createCustomLogger` since validation ensures `output_paths` is always set when needed

**Comment by ishaish103** [comment]
> default, this config

ive added default output paths to the defaultConfig but then it created a situation that UseInternalLogger is true by default (that was the value before) and output_path is set, which triggered validation error when running empty config:
` debug:
 
`
so i ended up removing this validation and added comment on the config param that in such case the output path will be ignored

**Comment by ishaish103** [comment]
> > so i ended up removing this validation and added comment on the config param that in such case the output path will be ignored
> 
> I'm worried that this will lead to multiple users of the Debug exporter setting `output_paths` in their configuration and pulling their hair wondering why it doesn't work. We should design software so that it's hard to make such mistakes.
> 
> How about not setting the `OutputPaths` field in `createDefaultConfig()` and only setting it in `createCustomLogger()` to `["stdout"]`, if it is unset (`nil`)? Would that work?

make sense, that worked!

File: exporter/debugexporter/factory.go [hunk 5]
```diff
@@ -48,7 +48,7 @@ func createDefaultConfig() component.Config {
 		SamplingInitial:    defaultSamplingInitial,
 		SamplingThereafter: defaultSamplingThereafter,
 		UseInternalLogger:  true,
-		QueueConfig:        configoptional.Default(exporterhelper.NewDefaultQueueConfig()),
+		QueueConfig:        configoptional.None[exporterhelper.QueueBatchConfig](),
```
Comment (andrzej-stencel) [inline exporter/debugexporter/factory.go]: Why is this change part of this pull request?

**Comment by ishaish103** [comment]
@andrzej-stencel  can take another look please?

File: exporter/debugexporter/metadata.yaml [hunk 6]
```diff
@@ -12,3 +12,4 @@ status:
     development: [profiles]
   distributions: [core, contrib, k8s]
   warnings: [Unstable Output Format]
+
```
Comment (andrzej-stencel) [inline exporter/debugexporter/metadata.yaml]: ? 🤷 😄

File: exporter/debugexporter/config_test.go [hunk 7]
```diff
@@ -53,10 +69,24 @@ func TestUnmarshalConfig(t *testing.T) {
 			factory := NewFactory()
 			cfg := factory.CreateDefaultConfig()
 			err = cm.Unmarshal(&cfg)
-			if tt.expectedErr != "" {
-				assert.ErrorContains(t, err, tt.expectedErr)
-			} else {
+			if err != nil {
+				if tt.expectedErr != "" {
```
Comment (andrzej-stencel) [inline exporter/debugexporter/config_test.go]: We should rather have separate fields in test cases for expected unmarshaling errors and expected validation errors. Can you change this?

File: exporter/debugexporter/README.md [hunk 8]
```diff
@@ -38,6 +38,7 @@ The following settings are optional:
   Refer to [Zap docs](https://godoc.org/go.uber.org/zap/zapcore#NewSampler) for more details
   on how sampling parameters impact number of messages.
 - `use_internal_logger` (default = `true`): uses the collector's internal logger for output. See [below](#using-the-collectors-internal-logger) for description.
+- `output_paths` (default = `["stdout"]`): a list of file paths to write logging output to. This option can only be used when `use_internal_logger` is `false`. Special strings "stdout" and "stderr" are interpreted as os.Stdout and os.Stderr respectively. All other values are treated as file paths. Setting `output_paths` when `use_internal_logger` is `true` results in a configuration error.
```
Comment (andrzej-stencel) [inline exporter/debugexporter/README.md]:
```suggestion
- `output_paths` (default = `["stdout"]`): a list of file paths to write output to. This option can only be used when `use_internal_logger` is `false`. Special strings "stdout" and "stderr" are interpreted as [standard output](https://en.wikipedia.org/wiki/Standard_streams#Standard_output_(stdout)) and [standard error](https://en.wikipedia.org/wiki/Standard_streams#Standard_error_(stderr)) respectively. All other values are treated as file paths. Setting `output_paths` when `use_internal_logger` is `true` results in a configuration error.
```

File: exporter/debugexporter/factory_test.go [hunk 9]
```diff
@@ -57,3 +70,280 @@ func TestCreateFactoryProfiles(t *testing.T) {
 	require.NoError(t, err)
 	assert.NotNil(t, te)
 }
+
+func TestCreateCustomLogger(t *testing.T) {
```
Comment (andrzej-stencel) [inline exporter/debugexporter/factory_test.go:72]: What is the purpose of this test?

File: exporter/debugexporter/factory_test.go [hunk 10]
```diff
@@ -57,3 +70,280 @@ func TestCreateFactoryProfiles(t *testing.T) {
...
+				// Ensure logger is fully synced and file handles are released
+				_ = logger.Sync()
+				time.Sleep(500 * time.Millisecond)
+				_ = logger.Sync()
+				time.Sleep(500 * time.Millisecond)
+				runtime.GC()
+				time.Sleep(200 * time.Millisecond)
+				// Now manually clean up the directory
+				cleanupDir()
+			}
```
Comment (andrzej-stencel) [inline exporter/debugexporter/factory_test.go]: The use of `time.Sleep` is a warning sign. Do we really need this?

File: exporter/debugexporter/factory_test.go [hunk 11]
```diff
@@ -57,3 +70,280 @@ func TestCreateFactoryProfiles(t *testing.T) {
...
+		})
+	}
+}
+
+func TestCreateLogger(t *testing.T) {
```
Comment (andrzej-stencel) [inline exporter/debugexporter/factory_test.go:137]: Same - why is this code needed?

**Comment by ishaish103** [comment]
Hey @andrzej-stencel  
i pushed fixes to your comments, please take another look

File: exporter/debugexporter/metadata.yaml (hunk 6 above)
Comment (ishaish103) [reply exporter/debugexporter/metadata.yaml]: Removed - was unintentional.

File: exporter/debugexporter/config_test.go (hunk 7 above)
Comment (ishaish103) [reply exporter/debugexporter/config_test.go]: Done - split into expectedUnmarshalErr and expectedValidateErr.

File: exporter/debugexporter/README.md (hunk 8 above)
Comment (ishaish103) [reply exporter/debugexporter/README.md]: Applied, thanks!

File: exporter/debugexporter/factory_test.go (hunk 9 above)
Comment (ishaish103) [reply exporter/debugexporter/factory_test.go:72]: Verifies createCustomLogger works with stdout, stderr, multiple paths, and nil (defaults to stdout).

File: exporter/debugexporter/factory_test.go (hunk 10 above)
Comment (ishaish103) [reply exporter/debugexporter/factory_test.go]: Removed - simplified tests by skipping file output test on Windows instead.

File: exporter/debugexporter/factory_test.go (hunk 11 above)
Comment (ishaish103) [reply exporter/debugexporter/factory_test.go:137]: Removed - was Windows cleanup workaround, now skipping that test on Windows.

File: exporter/debugexporter/factory.go (hunk 5 above)
Comment (ishaish103) [reply exporter/debugexporter/factory.go]: Reverted - was unintentional change.

**Review by andrzej-stencel — APPROVED** [review]
Thanks Ishai!

**Comment by andrzej-stencel** [comment]
I've removed the `release:risky-change` label added after commit https://github.com/open-telemetry/opentelemetry-collector/pull/14164/commits/a2ee4d02559b6dbf6dd8f232e5cb64cd95a6387f, because the Makefile change was reverted in https://github.com/open-telemetry/opentelemetry-collector/pull/14164/changes/1c1353a659550af4c0e0eeddc5ec5774cffe5541.

**Comment by andrzej-stencel** [comment]
Marked as `ready-to-merge` as it has codeowner's and approver's approvals.
